### ISSUE TITLE
🐞 OOM of patchcore and padim on Intel iGPU

### ISSUE THREAD

**Issue by zyz9740** [open-edge-platform/anomalib#2886]
### Describe the bug

When training padim and patchcore on Intel iGPU, out-of-memory occurs, killing the training process.

### Dataset

MVTecAD

### Model

PatchCore

### Steps to reproduce the behavior

pip install anomalib[full]==2.1
pip install torch torchvision  --index-url https://download.pytorch.org/whl/xpu --force-reinstall

```
import os
os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"

import argparse
from pathlib import Path
from anomalib.data import MVTecAD
from anomalib.models import Patchcore
from anomalib.engine import Engine
from anomalib.engine import Engine, SingleXPUStrategy, XPUAccelerator
from anomalib.data.utils import ValSplitMode


def arg_parser():
    parser = argparse.ArgumentParser(description='Train PatchCore model on MVTec bottle dataset', formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('-o', '--output', type=Path, default='ovmodels', required=False, help='Path to save the output OpenVINO IR.\n(default: %(default)s)')
    return parser.parse_args()

if __name__ == "__main__":
    args = arg_parser()
    
    print("Starting training PatchCore model on MVTec bottle dataset...")
    
    # Initialize the datamodule, model and engine for bottle category
    dataset_path = Path("../datasets/MVTecAD")

    datamodule = MVTecAD(
        root=dataset_path, 
        category="bottle", 
        train_batch_size=1, 
        eval_batch_size=1, 
        val_split_mode=ValSplitMode.FROM_TEST, 
        val_split_ratio=0.0001,
    )

    model = Patchcore()

    engine = Engine(
        strategy=SingleXPUStrategy(),
        accelerator=XPUAccelerator(),
    )
    
    # Train the model
    engine.fit(datamodule=datamodule, model=model)
    
    print("Training completed!")
    
```

### OS information

OS information:
- OS: ubuntu 22.04
- Python version: 3.12
- Anomalib version: 2.1.0 （v2.0.0 work well, no oom on both models）
- PyTorch version：2.8
- CUDA/cuDNN version: NA
- GPU models and configuration: 1270P iGPU
- Any other relevant information: 32G memory

### Expected behavior

Work well just as same as the behavior on v2.0.0.
I believe the 2 models should work well on MVTecAD dataset when training on Intel iGPU 

### Pip/GitHub

pip

### What version/branch did you use?

main

### Configuration YAML

```yaml
NA
```

### Logs

```shell
(anomalib_2_1) intel@intel-O-E-M:~/zyz/bmk_openvino/unsupervised/patchcore$ python train.py 
/home/intel/zyz/bmk_openvino/unsupervised/anomalib_2_1/lib/python3.12/site-packages/timm/models/layers/__init__.py:48: FutureWarning: Importing from timm.models.layers is deprecated, please import via timm.layers
  warnings.warn(f"Importing from {__name__} is deprecated, please import via timm.layers", FutureWarning)
Starting training PatchCore model on MVTec bottle dataset...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
Zero subset length encountered during splitting. This means one of your subsets
            might be empty or devoid of either normal or anomalous images.
Zero subset length encountered during splitting. This means one of your subsets
            might be empty or devoid of either normal or anomalous images.
/home/intel/zyz/bmk_openvino/unsupervised/anomalib_2_1/lib/python3.12/site-packages/lightning/pytorch/core/optimizer.py:183: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer

  | Name           | Type           | Params | Mode 
----------------------------------------------------------
0 | pre_processor  | PreProcessor   | 0      | train
1 | post_processor | PostProcessor  | 0      | train
2 | evaluator      | Evaluator      | 0      | train
3 | model          | PatchcoreModel | 24.9 M | train
----------------------------------------------------------
24.9 M    Trainable params
0         Non-trainable params
24.9 M    Total params
99.450    Total estimated model params size (MB)
19        Modules in train mode
174       Modules in eval mode
/home/intel/zyz/bmk_openvino/unsupervised/anomalib_2_1/lib/python3.12/site-packages/lightning/pytorch/utilities/data.py:106: Total length of `DataLoader` across ranks is zero. Please make sure this was your intention.
Epoch 0:  44%|████████████████████████████████████████████████████████████████████████████████████████████████████                                                                                                                             | 93/209 [00:10<00:12,  9.28it/s]Killed
Killed
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by rajeshgangireddy** [comment]
Hi @zyz9740 
We are currently testing anomalib models on iGPUs. While most models work out of the box for the newer iGPUs (from Lunar lake and Arrow lake), some models fail on iGPUs like Intel® Iris® Xe Graphics  (I believe this is the one you are using). 
However, 64 GB of shared memory is still good enough for these models to run. 
Moreover, for patchcore, OOM usually occurs in the next step (not the one where you are facing the error). 
Are all the drivers installed properly ? 

Please try the following :  
1) Could you check if torch is actually using xpu ? 
```
import torch 
torch.xpu.is_available() 
torch.xpu.get_device_properties()
```
2) Patchcore is memory bank based model. More images mean more memory required. Could you try with a smaller category of mvtec like "toothbrush" (this has only ~60 images). 

3) Can you send the output log with padim as well, if possible.

**Comment by zyz9740** [comment]
Thanks for your quick response !

For your first question.  In 1270P, the configuration is as following.

```
>>> import torch

>>> torch.xpu.is_available()

True
>>> torch.xpu.get_device_properties()
_XpuDeviceProperties(name='Intel(R) Iris(R) Xe Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', driver_version='1.6.31294+20', total_memory=29467MB, max_compute_units=96, gpu_eu_count=96, gpu_subslice_count=6, max_work_group_size=512, max_num_sub_groups=64, sub_group_sizes=[8 16 32], has_fp16=1, has_fp64=0, has_atomic64=1)

```

In addition, I have found another ARL-H device for testing, with the configuration listed below. I can run code on this device for the following discussion if needed.

OS information:

OS: windows 11
Python version: 3.13.5
Anomalib version: 2.1.0 
PyTorch version：2.8
CUDA/cuDNN version: NA
GPU models and configuration: ARL-H 265H
Any other relevant information: 32G memory * 2 = 64G

```
>>> import torch
... torch.xpu.is_available()
... torch.xpu.get_device_properties()
...
_XpuDeviceProperties(name='Intel(R) Arc(TM) 140T GPU (32GB)', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', driver_version='1.6.34177', total_memory=34376MB, max_compute_units=128, gpu_eu_count=128, gpu_subslice_count=16, max_work_group_size=1024, max_num_sub_groups=128, sub_group_sizes=[8 16 32], has_fp16=1, has_fp64=1, has_atomic64=1)

```

I make sure both the 2 devices use xpu because you can find that the GPU usage is nearly 100%

**Comment by zyz9740** [comment]
For second question. I don't think the memory consumption for xpu is reasonable, because whether I run the same code on anomalib v2.0.0 or on CPU. The memory consumption is only ~2G 

<img width="1478" height="220" alt="Image" src="https://github.com/user-attachments/assets/a3302ce8-fb78-4cd9-bbba-d6178a6f96c1" />


If I run it on xpu, it use nearly 30G and then is killed. 

<img width="1514" height="225" alt="Image" src="https://github.com/user-attachments/assets/38dce84e-01c9-4041-ae1e-0b0b2789920d" />


If it helps,  on 265H iGPU, it use nearly 40G memory whether running on CPU or iGPU. I'm not sure whether it is reasonable.

<img width="2045" height="622" alt="Image" src="https://github.com/user-attachments/assets/e22113f7-49a9-4846-a76a-0becd9386841" />


If I use a small datasets, it works well. But I don't understand why the memory consumption varies a lot on different devices and different version and what is corrent.

**Comment by zyz9740** [comment]
The log of padim is listed below. ( on Iris iGPU）

```
(anomalib_2_1) intel@intel-O-E-M:~/zyz/bmk_openvino/unsupervised/padim$ python train.py
/home/intel/zyz/bmk_openvino/unsupervised/anomalib_2_1/lib/python3.12/site-packages/timm/models/layers/__init__.py:48: FutureWarning: Importing from timm.models.layers is deprecated, please import via timm.layers
  warnings.warn(f"Importing from {__name__} is deprecated, please import via timm.layers", FutureWarning)
Starting training PaDiM model on MVTec bottle dataset...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
`Trainer(val_check_interval=1.0)` was configured so validation will run at the end of the training epoch..
Zero subset length encountered during splitting. This means one of your subsets
            might be empty or devoid of either normal or anomalous images.
Zero subset length encountered during splitting. This means one of your subsets
            might be empty or devoid of either normal or anomalous images.
/home/intel/zyz/bmk_openvino/unsupervised/anomalib_2_1/lib/python3.12/site-packages/lightning/pytorch/core/optimizer.py:183: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer

  | Name           | Type          | Params | Mode
---------------------------------------------------------
0 | pre_processor  | PreProcessor  | 0      | train
1 | post_processor | PostProcessor | 0      | train
2 | evaluator      | Evaluator     | 0      | train
3 | model          | PadimModel    | 2.8 M  | train
---------------------------------------------------------
2.8 M     Trainable params
0         Non-trainable params
2.8 M     Total params
11.131    Total estimated model params size (MB)
19        Modules in train mode
69        Modules in eval mode
/home/intel/zyz/bmk_openvino/unsupervised/anomalib_2_1/lib/python3.12/site-packages/lightning/pytorch/utilities/data.py:106: Total length of `DataLoader` across ranks is zero. Please make sure this was your intention.
Epoch 0:  92%|███████████████████████████████████████████████████████████████████████▋      | 192/209 [00:08<00:00, 21.98it/s]
Killed
```

<img width="1480" height="208" alt="Image" src="https://github.com/user-attachments/assets/3302e599-f522-4eab-a7e8-679247c7030e" />

Similar to patchcore, it works well on CPU

<img width="1480" height="208" alt="Image" src="https://github.com/user-attachments/assets/67764481-2983-4d0d-872c-6010224f126f" />





On 265H， CPU
<img width="2099" height="608" alt="Image" src="https://github.com/user-attachments/assets/9e9b8e39-01a5-4e5f-98f5-6dbea7c8bd62" />


On 265H，iGPU

<img width="2070" height="594" alt="Image" src="https://github.com/user-attachments/assets/b737b58c-df3b-4d42-b5ff-b423a5c3f81f" />

**Comment by zyz9740** [comment]
For this question. I find that torch xpu module doesn't formally support Iris iGPU so I can use Arc iGPU for reproduction. But I think it is a common problem on both patchore and padim on iGPU（whether the iGPU is Iris or Arc）

Besides, I only reproduced it on the latest version v2.1.0 instead of v2.0.0 so that I choose to use the older version now. I'm not sure whether this problem is related to https://github.com/open-edge-platform/anomalib/pull/2813 merged in v2.1.0

**Comment by rajeshgangireddy** [comment]
@zyz9740 
Thank you so much for the detailed information. This is helpful. 


> I find that torch xpu module doesn't formally support Iris iGPU

This is true, even if the training starts, some of the operations might not be supported on Iris iGPU
For example, https://github.com/pytorch/pytorch/issues/149953#issuecomment-3164223547
From what i gather, the intel-pytorch team is actively fixing this. 




> Besides, I only reproduced it on the latest version v2.1.0 instead of v2.0.0 so that I choose to use the older version now. I'm not sure whether this problem is related to [#2813](https://github.com/open-edge-platform/anomalib/pull/2813) merged in v2.1.0

This is interesting. When we made those changes, we haven't tested them on iGPU. But now, as we want to support anomalib on xpu as well, I will look into it and see if we can fix it.

**Comment by DeepKnowledge1** [comment]
@zyz9740 

That’s a common issue with Intel iGPUs — memory gets maxed out fast when training PaDiM.

In AnomaVision we solved this by:


https://github.com/DeepKnowledge1/AnomaVision


**Cutting memory usage by 40–60% ⚡**

Adding smart memory management that lets you train on datasets up to 5× larger without OOM

Providing an `evaluate_memory_efficient` mode specifically for resource-constrained devices

If you’re running into OOM on Intel GPUs, this approach could help stabilize training and keep it running smoothly.


**Note:** 
We only support PaDim

**Comment by alfieroddan** [comment]
@zyz9740 Can confirm your results using NVIDIA GPU as well...something clearly not right.

v2.0.0

<img width="1337" height="583" alt="Image" src="https://github.com/user-attachments/assets/b55b8773-8415-4833-a8b3-eb7b3b443094" />

v2.1.0

<img width="1215" height="562" alt="Image" src="https://github.com/user-attachments/assets/0bcad410-cd1a-43ce-a3f6-65d2c8dd6357" />

The PR #2813 was supposed to the opposite, will look into this 🙁

**Comment by alfieroddan** [comment]
Just to update everyone here, there's two things happening here.

#1 Pre or Post Processing

In commit 03dc12694519b9d4d3af1dc3987062388abd6040 we see an increase in reserved memory:

<img width="1184" height="473" alt="Image" src="https://github.com/user-attachments/assets/1113de84-22d2-4514-b21b-b16570e11c32" />

#2 Cat vs List Allocation

In PR #2813 we build the bank with `new_bank = torch.cat((self.memory_bank, embedding), dim=0).to(self.memory_bank)`. But in PyTorch, doing torch.cat repeatedly doesn't free memory (works better in toy examples where you only do it once). PyTorch keeps old memory blocks to optimize performance, which can lead to high memory usage over time.

To avoid this, it's better to collect embeddings in a list and use torch.cat once, or pre-allocate the memory bank if the final size is known. Example with using list instead of cat:

<img width="1184" height="488" alt="Image" src="https://github.com/user-attachments/assets/0668a0d7-2e1f-4831-8332-d92dc63e19b7" />

Notice the drop in allocated memory which PR #2813 was trying to solve 😄 .

Will make a PR to change from cat to list, then we can discuss the pre-post processing at a later date.
